include/cuco/detail/static_set/static_set_ref.inl

```diff
@@ -187,68 +246,137 @@ class operator_impl<op::insert_tag,
-        } else if constexpr (thread_scope == cuda::thread_scope_block) {
-          return atomicCAS_block(reinterpret_cast<unsigned int*>(slot), *expected_ptr, *value_ptr);
-        } else {
-          static_assert(cuco::dependent_false<decltype(thread_scope)>, "Unsupported thread scope");
+    using insert_result = typename ref_type::insert_result;
+
+    ref_type& ref_    = static_cast<ref_type&>(*this);
+    auto probing_iter = ref_.probing_scheme_(value, ref_.storage_ref_.num_windows());
+
+    while (true) {
+      auto const window_slots = ref_.storage_ref_[*probing_iter];
+
+      for (auto i = 0; i < window_size; ++i) {
+        auto const eq_res      = ref_.predicate_(window_slots[i], value);
+        auto const* window_ptr = (ref_.storage_ref_.data() + *probing_iter)->data();
+
+        // If the key is already in the container, return false
+        if (eq_res == detail::equal_result::EQUAL) { return {iterator{&window_ptr[i]}, false}; }
+        if (eq_res == detail::equal_result::EMPTY) {
+          switch (ref_.attempt_insert(window_ptr + i, value)) {
+            case insert_result::SUCCESS: {
+              return {iterator{&window_ptr[i]}, true};
+            }
+            case insert_result::DUPLICATE: {
+              return {iterator{&window_ptr[i]}, false};
+            }
+            default: continue;
+          }
+        }
+      }
+      ++probing_iter;
+    };
+  }
+
+  /**
+   * @brief Inserts the given element into the set.
+   *
+   * @note This API returns a pair consisting of an iterator to the inserted element (or to the
+   * element that prevented the insertion) and a `bool` denoting whether the insertion took place or
+   * not.
+   *
+   * @param group The Cooperative Group used to perform group insert_and_find
+   * @param value The element to insert
+   *
+   * @return a pair consisting of an iterator to the element and a bool indicating whether the
+   * insertion is successful or not.
+   */
+  __device__ thrust::pair<iterator, bool> insert_and_find(
+    cooperative_groups::thread_block_tile<cg_size> const& group, value_type const& value) noexcept
+  {
+    using insert_result = typename ref_type::insert_result;
+
+    ref_type& ref_    = static_cast<ref_type&>(*this);
+    auto probing_iter = ref_.probing_scheme_(group, value, ref_.storage_ref_.num_windows());
+
+    while (true) {
+      auto const window_slots = ref_.storage_ref_[*probing_iter];
+
+      auto const [state, intra_window_index] = [&]() {
+        for (auto i = 0; i < window_size; ++i) {
+          switch (ref_.predicate_(window_slots[i], value)) {
+            case detail::equal_result::EMPTY: return cuco::pair{detail::equal_result::EMPTY, i};
+            case detail::equal_result::EQUAL: return cuco::pair{detail::equal_result::EQUAL, i};
+            default: continue;
+          }
         }
+        // returns dummy index `-1` for UNEQUAL
+        return cuco::pair<detail::equal_result, int32_t>{detail::equal_result::UNEQUAL, -1};
```

Comment:
we could use a struct to get rid of `cuco::pair` here

Reply:
I think we can leave it as is. No need to define a struct to make a single line of code/implementation detail look nicer.